rust/pytests/rootcanal.rs

```diff
@@ -0,0 +1,295 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use anyhow::anyhow;
+use cfg_if::cfg_if;
+use log::{debug, info, warn};
+use rand::Rng;
+use std::{env, fs, future::Future, io::Read as _, path};
+use tokio::{
+    io::{self, AsyncBufReadExt as _},
+    select,
+    sync::mpsc,
+    task,
+};
+use tokio_util::sync::CancellationToken;
+
+/// Run the provided closure with rootcanal running in the background.
+///
+/// Port info for the rootcanal instance is passed to the closure; those ports must be used rather
+/// than the default rootcanal ports so that multiple tests may run concurrently without clashing.
+///
+/// Sets up `env_logger`, if it wasn't already configured, so that debug logging with rootcanal
+/// output can be enabled with `RUST_LOG=debug` env var.
+///
+/// If the correct rootcanal binary isn't already cached, it will be downloaded from GitHub
+/// Releases if a binary is available for the current OS & architecture.
+///
+/// This is a stop-gap until rootcanal's build is improved to the point that a rootcanal-sys crate,
+/// and using rootcanal as a library rather than a separate process, becomes feasible.
+pub(crate) async fn run_with_rootcanal<O, F>(closure: F) -> anyhow::Result<()>
```

Comment:
@hchataing forgive me for these rootcanal crimes. Let me know if there is a less bad way to do this while I try to get the rootcanal build working on more systems.